Couchforms analytics

dannyroberts opened the pull request:
@czue waiting for tests

There's lots more of these to go, but probably no harm merging this in first

dannyroberts committed:
add domain_has_submission_in_last_30_days analytics function

dannyroberts committed:
add stale=ok to domain_has_submission_in_last_30_days

dannyroberts committed:
add get_number_of_forms_per_domain

dannyroberts committed:
add get_number_of_forms_in_domain

dannyroberts committed:
make domain.calculations.active use domain_has_submission_in_last_30_days

dannyroberts committed:
add get_{first,last}_form_submission_received

dannyroberts committed:
have domain_has_submission_in_last_30_days use get_last_form_submission_received

dannyroberts committed:
remove active_users hqadmin view

it times out on prod anyway

dannyroberts committed:
rewrite serverup couch_check to use a non-form/case view

dannyroberts committed:
add get_form_ids_for_user and get_form_xml_element dbaccessors

dannyroberts committed:
pull out app_has_been_submitted_to_in_last_30_days

and make that query use limit=1!

dannyroberts committed:
remove unused function

czue commented:
looks like a great set of changes. I think I'll feel much more comfortable if we have tests that exercise these functions. once those exist :shipit:

dannyroberts committed:
add couchforms.analytics tests

dimagimon commented on corehq/ex-submodules/couchforms/analytics.py:
```diff
@@ -0,0 +1,105 @@
+            group=True,
+            group_level=2,
+            startkey=key,
+            endkey=key+[{}],
```
E226 missing whitespace around arithmetic operator

dimagimon commented on corehq/ex-submodules/couchforms/analytics.py:
```diff
@@ -0,0 +1,105 @@
+    row = XFormInstance.get_db().view(
+        "reports_forms/all_forms",
+        startkey=key,
+        endkey=key+[{}],
```
E226 missing whitespace around arithmetic operator

dannyroberts committed:
remove old cleanup mgmt cmd

that referenced reports_forms/all_forms

dimagimon commented on corehq/ex-submodules/couchforms/tests/test_analytics.py:
```diff
@@ -0,0 +1,62 @@
+
+    @classmethod
+    def setUpClass(cls):
+        from casexml.apps.case.tests import delete_all_xforms
```
F401 'delete_all_xforms' imported but unused

dannyroberts committed:
remove historical_forms

dimagimon commented on corehq/apps/ota/views.py:
```diff
@@ -14,6 +14,7 @@
 from corehq.apps.style.views import BaseB3SectionPageView
 from corehq.apps.users.models import CouchUser, CommCareUser
 from corehq.util.view_utils import json_error
+from couchforms.dbaccessors import get_form_ids_for_user, get_form_xml_element
```
F401 'get_form_ids_for_user' imported but unused

dimagimon replied:
F401 'get_form_xml_element' imported but unused

dannyroberts committed:
move get_commtrack_forms to dbaccessors

dimagimon commented:
There are 1 errors:

* corehq/apps/commtrack/tests/util.py, line 12 - F401 'get_commtrack_forms' imported but unused

dannyroberts committed:
pull couchforms analytics out of corehq/apps/reports/util.py

- get_username_in_last_form_user_id_submitted
- get_all_user_ids_submitted
- tests for both
- small functional change to get_username_from_forms:
  no longer excludes username "none" from form

dimagimon commented:
There are 4 errors:

* corehq/apps/commtrack/tests/util.py, line 12 - F401 'get_commtrack_forms' imported but unused
* corehq/ex-submodules/couchforms/analytics.py, line 2 - F401 'get_db' imported but unused
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 26 - E501 line too long (125 > 115 characters)
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 28 - E501 line too long (123 > 115 characters)

dannyroberts committed:
rewrite get_all_user_ids_submitted to use reports_forms/all_forms

- no more references to reports_forms/all_submitted_users

dimagimon commented:
There are 4 errors:

* corehq/apps/commtrack/tests/util.py, line 12 - F401 'get_commtrack_forms' imported but unused
* corehq/ex-submodules/couchforms/analytics.py, line 2 - F401 'get_db' imported but unused
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 26 - E501 line too long (125 > 115 characters)
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 28 - E501 line too long (123 > 115 characters)

dannyroberts committed:
Merge branch 'master' into couchforms-analytics

dannyroberts committed:
Merge branch 'master' into couchforms-analytics

dimagimon commented:
There are 4 errors:

* corehq/apps/commtrack/tests/util.py, line 12 - F401 'get_commtrack_forms' imported but unused
* corehq/ex-submodules/couchforms/analytics.py, line 2 - F401 'get_db' imported but unused
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 26 - E501 line too long (125 > 115 characters)
* corehq/ex-submodules/couchforms/tests/test_analytics.py, line 28 - E501 line too long (123 > 115 characters)

dannyroberts committed:
pep8

dannyroberts committed:
make datespan_from_beginning use get_first_form_submission_received

dannyroberts committed:
add and test get_all_xmlns_app_id_pairs_submitted_to_in_domain

dannyroberts committed:
add and test get_last_form_submission_for_user_for_app

dimagimon commented on corehq/ex-submodules/couchforms/analytics.py:
```diff
@@ -0,0 +1,162 @@
+        key = ['submission user', domain, user_id]
+    xform = XFormInstance.view(
+        "reports_forms/all_forms",
+        startkey=key+[{}],
```
E226 missing whitespace around arithmetic operator

dannyroberts committed:
add and test get_number_of_submissions

dannyroberts committed:
remove unused patch_submissions mgmt cmd

dannyroberts committed:
make ptop_fast_reindex_reportxforms use couchforms/all_submissions_by_domain

removing last reference to hqadmin/domains_over_time

dannyroberts commented:
:white_check_mark:

dannyroberts commented:
@czue thanks for suggesting tests. Adding them was super easy and super helpful.

czue commented:
looks like a great set of changes. reviewed individual commits for structure though didn't review the functions closely. I think I'm comfortable with that level given that it was mostly moved code and that it now has test coverage too (!!). will merge assuming you agree with that sentiment.

one question - wanted to confirm that you were careful about the convention of `analytics` being something that is only used in reports? so like, we can now rewrite those functions on ES or sofabed with no major concerns. right?

assuming yes is probably worth highlighting the convention to the wider team

dannyroberts commented:
@czue yeah, that's right.

@dimagi/team-commcare-hq I'm setting a new convention in this PR, most relevant to our bulkier data models i.e. forms & cases, of splitting out `analytics` accessors from `dbaccessors`. The idea is that `dbaccessors` are what's needed for core operations and must be always up-to-date and exactly correct, whereas `analytics` can and should use `stale=stale_ok()` if querying couch or can be rewritten to work off ES or sofabed, because they're used for reports and analytics and other read-heavy applications where exact up-to-dateness is not a hard requirement.

dannyroberts commented:
As a bonus, this PR removes the last usages of the following couch views:

- `reports_forms/all_submitted_users`
- `hqadmin/domains_over_time`

which we can remove at any time here: https://github.com/dimagi/commcare-hq/pull/8341 (but will require a pretty large reindex)
